### ISSUE TITLE
Proposal to allow extension of reactive forms

### ISSUE THREAD

**Issue by norganos** [angular/angular#32004]
# 🚀 feature request

### Relevant Package
This feature request is for @angular/forms

### Description
I want to minimize repetition of similar code related to forms and re-use the reactive form connection (formControl bound to input component via reactive forms directives) for other states than value/disabled.
Developers already are able to implement custom ControlValueAccessors, so one approach would be to override FormControl class and provide necessary properties there. The problem is, that a custom ControlValueAccessor does not get the bound FormControl directly. Existing implementation only connects value and disabled state.

For Example:
* a custom SelectControlValueAccessor could automatically generate option tags inside select tag if bound formControl has a "elements" property.
* FormControls for numeric values with a informational property describing the measurement unit -> a custom NumberControlValueAccessor could automatically place a postfix tag or a custom component could use this unit.
* a custom component showing errors of formControl automatically.
* custom FormControl objects that not only have a "disabled" state but also a reason why it's disabled => custom components/controlValueAccessors could display these reason automatically for better user experience
Having additional data besides value and disabled on the FormControl objects has the advantage that it integrates nicely when such data comes from the same backend as the values themselves.

### Describe the solution you'd like
The simplest approach is to provide a hook mechanism to get some custom code be called whenever setUpControl and cleanUpControl from forms/shared.ts are invoked.
This approach is implemented in [norganos/angular:reactive_forms_hooks](https://github.com/norganos/angular/tree/reactive_forms_hooks)

#### Example usage
The following code example shows how these hooks could be used to transport a unit postfix.

##### Custom Input Component
For simplicity, we create a component that wraps a number input and the unit
```typescript
@Component({
  selector: 'app-measurement',
  template: `<input type="number" name="custom" [(ngModel)]="model" (ngModelChange)="changeFn($event)" [disabled]="isDisabled"><span>{{unit }}</span>`,
  providers: [{provide: NG_VALUE_ACCESSOR, multi: true, useExisting: MeasurementInput}]
})
class MeasurementInput implements ControlValueAccessor {
  model = 0;
  @Input('disabled') isDisabled: boolean = false;
  changeFn: (value: any) => void;
  unit = '';
  writeValue(value: any) { this.model = value; }
  registerOnChange(fn: (value: any) => void) { this.changeFn = fn; }
  registerOnTouched() {}
  setDisabledState(isDisabled: boolean) { this.isDisabled = isDisabled; }
  setUnit(unit: string) { this.unit = unit};
}
```

##### Custom FormControl
Of course we need a place to store the unit. For simplicity, we just define a single "unit" property here.
```typescript
class MeasurementFormControl extends FormControl {
  private _onUnitChange: Function[] = [];
  private _unit = '';
  get unit(): string { return this._unit; }
  set unit(value: string) {
    this._unit = value;
    for (let f of this._onUnitChange) { f(value); }
  }
  constructor(formState: any, unit: string, validatorOrOpts?: ValidatorFn | ValidatorFn[] | AbstractControlOptions | null, asyncValidator?: AsyncValidatorFn | AsyncValidatorFn[] | null) {
    super(formState, validatorOrOpts, asyncValidator);
    this.unit = unit;
  }
  registerOnUnitChange(fn: (unit: string) => void): void {
    this._onUnitChange.push(fn);
  }
  clearUnitChangeFunctions(): void {
    this._onUnitChange = [];
  }
}
```

##### Hooks
This example hard codes MeasurementInput and MeasurementFormControl. a real world implementation of course would use interfaces and type guards.
```typescript
@Injectable()
class MeasurementFormsHook implements FormsHook {
  setUpControl(control: FormControl, dir: NgControl): void {
    const accessor = dir.valueAccessor;
    if (accessor != null && (<MeasurementInput>accessor).setUnit !== undefined&& (<MeasurementFormControl>control).registerOnUnitChange !== undefined) {
      control.registerOnUnitChange(unit=> accessor.setUnit(unit));
      accessor.setUnit(control.unit);
    }
  },
  cleanUpControl(control: FormControl, dir: NgControl): void {
    if ((<MeasurementFormControl>control).clearUnitChangeFunctions !== undefined) {
      control.clearUnitChangeFunctions();
    }
  }
}
```

##### Register Hooks
```typescript
@NgModule({
  imports: [
    ReactiveFormsModule
  ],
  providers: [
    {provide: NG_FORMS_HOOK, useClass: MeasurementFormsHook}
  ]
})
class AppModule {}
```

##### Usage
A component then can use the custom components and controls
```typescript
@Component({
  template: `
  <form [formGroup]="form">
    <app-measurement formControlName="width"></app-measurement>
    <app-measurement formControlName="height"></app-measurement>
    <app-measurement formControlName="weight"></app-measurement>
  </form>`
})
class FormGroupNameComp implements OnInit {
  form: FormGroup;
  ngOnInit() {
    this.form = new FormGroup({
      width: new MeasurementFormControl(12, 'm'),
      height: new MeasurementFormControl(50, 'cm'),
      weight: new MeasurementFormControl(23, 'kg'),
    })
  }
}
```

### Describe alternatives you've considered
Issue #31963 could also address these possibilities, but it looks like it's a bigger thing.

The proposal from issue #19686 (a general purpose meta-data container on AbstractControl) should integrate nicely with this approach here.

**Comment by Airblader** [comment]
As a library author I would really like avoiding mixing the CVA infrastructure with forms directly. As much as there is wrong at the moment, this separation allows libraries to reuse the infrastructure most component libraries ship with which is a clear benefit.

**Comment by norganos** [comment]
Well, my intent is not to mix them directly.
Perhaps it is clearer if I explain my use case in more detail:
- Although the form structure remains (mostly) stable in our application, details (i.e. meta data such as constraints, enums, units, etc) about every field is provided by the backend and it can change (as the backend itself is getting these value from an even more back backend ;-))
- So, the naive approach is to retrieve the meta-data and pass it along side to the form in some other properties
- For components whose formGroups contain 30+ controls this adds up to a complete mess. in the typescript files and the templates
- most of the CVAs (some native inputs, mostly angular components) can consume the metadata etc by custom input properties.
- My proposal just enables the developer (me) to have the "take that from meta-data and put it into this input-property" happen in the background for all controls and not having to repeat it every time.
- I would never expect a library to ship or register a hook here

**Comment by Airblader** [comment]
I don't have a better solution to offer right now (sorry :-)), but I still see conceptual problems with the proposal, for example this:

```
(<MeasurementInput>accessor).setUnit
```

This assumes that the CVA is implemented together with the component, which in your case is true, but in general this is false, in fact the CVAs shipped with Angular itself are a prime example of providing the accessor from the "outside". 

I realize that since you're not changing the CVA contract the argument is "it's optional, so it doesn't matter", but it showcases the implicit — and incorrect — assumption that the CVA is an abstract version of the component; in reality it is a contract describing how to read and write a value to/from the component.

This isn't something you fix with type guards, but something you'd workaround using type *assertions*, simply tricking the TS compiler.

**Comment by Airblader** [comment]
I wonder if we couldn't extend NgControl with some hook that the setupControl etc call. I think NgControl is a more appropriate access to the underlying component than the CVA.

**Comment by norganos** [comment]
That's how I ended up here ;)
I started by extending FormControlto have a suitable place where all data belonging to one field is stored.
Then I wanted to somehow get the things into the input elements/components.
Turned out, at least formControlName doesn't get a hold of the CVA directly (and that's our usual case), but the setUpControl from shared.ts does the magic.
I ended up extending all reactive forms directives (with the team having to use our formXY directives instead of the usual ones; a big bug source imho) as well as copying lots of code from the shared.ts and although I finished it, it felt wrong from the beginning, of course.
Just by being able to register a hook, I'd be able to achieve the result in less than hundred LoCs...
And please don't fixate on the example in the issue, I wanted to show how the hook is registered and when it's called without actually doing a clean complete example.
But actually, as you mention it, there's a case I haven't covered: as angular expects only one provided custom CVA, we could not replace the CVA implementation from a library component.

**Comment by thefliik** [comment]
@norganos not to invalidate your request, but recently I've proposed a (breaking) update to `AbstractControl` (intended to be part of a hypothetical `ReactiveFormsModule2`) #31963.

At the end of the proposal, I present a new `ControlValueAccessor` API which would address the need you describe here in (I think) a more flexible way (though again, as a breaking change). I'd be interested to hear if that update to `ControlValueAccessor` would address your use case.

**Comment by thefliik** [comment]
Having read through this proposal more closely, it appears that the goals described here can currently be accomplished by providing a custom `FormControlDirective` which provides special considerations for a custom `FormControl`. It's unclear to me why this proposal is superior to simply extending `FormControlDirective` and `ControlValueAccessor`?

For example, without delving into too much detail, you should be able to accomplish the following (unless I'm misunderstanding something):

```ts
class MeasurementFormControl extends FormControl {
  // ...
}

export class CustomFormControlDirective extends FormControlDirective {
  // provide special considerations for MeasurementFormControl
}

@NgModule({
  imports: [
    ReactiveFormsModule
  ],
  providers: [
    {provide: FormControlDirective, useClass: CustomFormControlDirective}
  ]
})
class AppModule {}
```

Admittedly, the `FormControlDirective` implementation is surprisingly complex, making any extensions to it more difficult, but that seems like a separate issue.

**Comment by norganos** [comment]
@thefliik I've seen your proposal yesterday morning and was surprised by the coincidence. Yet I didn't want to throw mine into the thrash ;-)
I think a ReactiveFormsModule2 would indeed solve most of my problems 👍 

As you said, a refactoring of FormControlDirective with the shared code being overridable would make it way easier for this as well.

It's totally correct, that my hooks proposal is more a consequence of code being too closed in the first place (just like back then in Java Swing where almost everything was declared final)

**Comment by thefliik** [comment]
> `@thefliik` I've seen your proposal yesterday morning and was surprised by the coincidence. Yet I didn't want to throw mine into the thrash ;-)

For sure! And again, since my proposal is a breaking change, there is definitely room for both improvements. When I saw this issue though, it seemed important to confirm that my proposal would also address this use case (as long as we're "fixing" ReactiveFormsModule, it better be a *comprehensive* fix).

> It's totally correct, that my hooks proposal is more a consequence of code being too closed in the first place. I just thought, a minimal invasive change would probably be better.

That's fair. And I don't have a great counter proposal for you (at least, one that's also non-breaking). Even moving the shared code into the classes (making it overridable), there is still the problem that you (potentially) need to have a deep understanding of the implementation to override it effectively, so I can see the merit of this proposal.

I guess my (perhaps unfounded) fear is that maybe this proposal only works for a slice of the potential customizations people want, and so other folks will still need to fall back on extending `FormControlDirective`. In which case a (non-breaking) improvement to the `FormControlDirective` API is still needed (and maybe that improvement would also largely address this issue).

**Comment by norganos** [comment]
So, thanks to @thefliik and @Airblader, I came up with a different approach for my requirement, which I wanted to share in case anyone has similar requirements:

- instead of trying to hook into the existing reactive form methods, I now snuggle with it ;-)
- I wrote a interface similar to CVA, that basically acts as the receiving part of my extended formControl metadata (I called it ControlMetaAccessor [=CMA in the following] and it offers methods like setConstraints which accepts a struct containing all kind of constraints. I also use that struct in the extended formControl to automatically create validators out of it)
- and a injectorToken for CMA just like NG_VALUE_ACCESSOR is for CVA
- Then I created 2 directives that use the same selectors formControl and formControlName. these directives get a hold of a CMA (out of the injector, just like CVAs) and they get the bound formControl (there I had to copy around 10 lines of the reactive forms code: formControlName directive has to get the control by path from the closest formGroup)
- if the directives find a CMA and the formControl is a extended formControl then they set up the connection.
- on top I wrote some default CMA directives (just like angular does with the builtin CVAs) to enable CMA functionality for html inputs and some library components

This works like a charm and it feels much cleaner than the proposed hook solution.
So, very thanks for the discussion!
Should I close this issue then?
